client.cfg

```
@@ -291,4 +291,23 @@ gcode:
   {% set speed = params.SPEED|default(client.speed_retract)|default(35) %}
 
   _CLIENT_EXTRUDE LENGTH=-{length|float|abs} SPEED={speed|float|abs}
-  
+
+[gcode_macro _CLIENT_LINEAR_MOVE]
+description: Linear move with save and restore of the gcode state
+gcode:
+  {% set x_move = "X" ~ params.X if params.X is defined else "" %}
+  {% set y_move = "Y" ~ params.Y if params.Y is defined else "" %}
+  {% set z_move = "Z" ~ params.Z if params.Z is defined else "" %}
+  {% set e_move = "E" ~ params.E if params.E is defined else "" %}
+  {% set rate = "F" ~ params.F if params.F is defined else "" %}
+  {% set ABSOLUTE = params.ABSOLUTE | default(0) | int != 0 %}
+  {% set ABSOLUTE_E = params.ABSOLUTE_E | default(0) | int != 0 %}
+  SAVE_GCODE_STATE NAME=_client_movement
+  {% if 'E' in params %}
+    M8{ 2 if ABSOLUTE_E else 3 }
+  {% endif %}
+  {% if 'X' in params or 'Y' in params or 'Z' in params %}
+    G9{ 0 if ABSOLUTE else 1 }
+  {% endif %}
```

I think Klipper doesn't work like that, but [in Marlin](https://marlinfw.org/docs/gcode/G090.html), `G90` and `G91` also sets the extruder positioning mode, so I think we should change the order of these.

```suggestion
  {% if 'X' in params or 'Y' in params or 'Z' in params %}
    G9{ 0 if ABSOLUTE else 1 }
  {% endif %}
  {% if 'E' in params %}
    M8{ 2 if ABSOLUTE_E else 3 }
  {% endif %}
```